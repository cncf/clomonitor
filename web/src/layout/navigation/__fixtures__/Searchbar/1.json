{
  "items": [
    {
      "category_id": 0,
      "description": "Artifact Hub is a web-based application that enables finding, installing, and publishing packages and configurations for CNCF projects.",
      "devstats_url": "https://artifacthub.devstats.cncf.io/",
      "display_name": "Artifact Hub",
      "id": "00000000-0001-0000-0000-000000000000",
      "logo_url": "https://raw.githubusercontent.com/cncf/artwork/master/projects/artifacthub/icon/color/artifacthub-icon-color.svg",
      "maturity_id": 2,
      "name": "artifact-hub",
      "organization": {
        "name": "artifact-hub"
      },
      "rating": "a",
      "repositories": [
        {
          "kind": "primary",
          "name": "artifact-hub",
          "url": "https://github.com/artifacthub/hub"
        }
      ],
      "score": {
        "best_practices": 100,
        "documentation": 80,
        "global": 95,
        "license": 100,
        "score_kind": "Primary",
        "security": 100
      },
      "updated_at": 1645534816
    },
    {
      "category_id": 0,
      "description": "Backstage is an open platform for building developer portals.",
      "devstats_url": "https://backstage.devstats.cncf.io/",
      "display_name": "Backstage",
      "id": "00000000-0046-0000-0000-000000000000",
      "logo_url": "https://raw.githubusercontent.com/cncf/artwork/master/projects/backstage/icon/color/backstage-icon-color.svg",
      "maturity_id": 2,
      "name": "backstage",
      "organization": {
        "name": "backstage"
      },
      "rating": "a",
      "repositories": [
        {
          "kind": "primary",
          "name": "backstage",
          "url": "https://github.com/backstage/backstage"
        }
      ],
      "score": {
        "best_practices": 35,
        "documentation": 95,
        "global": 78,
        "license": 80,
        "score_kind": "Primary",
        "security": 100
      },
      "updated_at": 1645606822
    },
    {
      "category_id": 0,
      "description": "Cloud-native application life-cycle orchestration. Keptn automates your SLO-driven multi-stage delivery and operations & remediation of your applications.",
      "devstats_url": "https://keptn.devstats.cncf.io/",
      "display_name": "Keptn",
      "id": "00000000-0067-0000-0000-000000000000",
      "logo_url": "https://raw.githubusercontent.com/cncf/artwork/master/projects/keptn/icon/color/keptn-icon-color.svg",
      "maturity_id": 2,
      "name": "keptn",
      "organization": {
        "name": "keptn"
      },
      "rating": "a",
      "repositories": [
        {
          "kind": "primary",
          "name": "keptn",
          "url": "https://github.com/keptn/keptn"
        }
      ],
      "score": {
        "best_practices": 95,
        "documentation": 95,
        "global": 93,
        "license": 80,
        "score_kind": "Primary",
        "security": 100
      },
      "updated_at": 1645606817
    },
    {
      "category_id": 0,
      "description": "Make shipping applications more enjoyable.",
      "devstats_url": "https://kubevela.devstats.cncf.io/",
      "display_name": "KubeVela",
      "home_url": "https://kubevela.io/",
      "id": "00000000-0072-0000-0000-000000000000",
      "logo_url": "https://raw.githubusercontent.com/cncf/artwork/master/projects/kubevela/icon/color/kubevela-icon-color.svg",
      "maturity_id": 2,
      "name": "kubevela",
      "organization": {
        "name": "oam-dev"
      },
      "rating": "a",
      "repositories": [
        {
          "kind": "secondary",
          "name": "spec",
          "url": "https://github.com/oam-dev/spec"
        },
        {
          "kind": "primary",
          "name": "kubevela",
          "url": "https://github.com/oam-dev/kubevela"
        }
      ],
      "score": {
        "best_practices": 100,
        "documentation": 84,
        "global": 82,
        "license": 64,
        "score_kind": "Primary",
        "security": 100
      },
      "updated_at": 1645599620
    },
    {
      "category_id": 0,
      "description": "Kubernetes Virtualization API and runtime in order to define and manage virtual machines.",
      "devstats_url": "https://kubevirt.devstats.cncf.io/",
      "display_name": "KubeVirt",
      "id": "00000000-0073-0000-0000-000000000000",
      "logo_url": "https://raw.githubusercontent.com/cncf/artwork/master/projects/kubevirt/icon/color/kubevirt-icon-color.svg",
      "maturity_id": 2,
      "name": "kubevirt",
      "organization": {
        "name": "kubevirt"
      },
      "rating": "a",
      "repositories": [
        {
          "kind": "primary",
          "name": "kubevirt",
          "url": "https://github.com/kubevirt/kubevirt"
        }
      ],
      "score": {
        "best_practices": 70,
        "documentation": 80,
        "global": 88,
        "license": 100,
        "score_kind": "Primary",
        "security": 100
      },
      "updated_at": 1645599617
    },
    {
      "category_id": 0,
      "description": "Automate application management on Kubernetes.",
      "devstats_url": "https://openkruise.devstats.cncf.io/",
      "display_name": "OpenKruise",
      "id": "00000000-0085-0000-0000-000000000000",
      "logo_url": "https://raw.githubusercontent.com/cncf/artwork/master/projects/openkruise/icon/color/openkruise-icon-color.svg",
      "maturity_id": 2,
      "name": "open-kruise",
      "organization": {
        "name": "open-kruise"
      },
      "rating": "a",
      "repositories": [
        {
          "kind": "primary",
          "name": "kruise",
          "url": "https://github.com/openkruise/kruise"
        }
      ],
      "score": {
        "best_practices": 95,
        "documentation": 90,
        "global": 91,
        "license": 80,
        "score_kind": "Primary",
        "security": 100
      },
      "updated_at": 1645524719
    },
    {
      "category_id": 0,
      "description": "An early-stage event processing system for unstructured data with rich support for structural pattern-matching, filtering and transformation.",
      "devstats_url": "https://tremor.devstats.cncf.io/",
      "display_name": "Tremor",
      "id": "00000000-0104-0000-0000-000000000000",
      "logo_url": "https://raw.githubusercontent.com/cncf/artwork/master/projects/tremor/icon/color/tremor-icon-color.svg",
      "maturity_id": 2,
      "name": "tremor",
      "organization": {
        "name": "tremor"
      },
      "rating": "a",
      "repositories": [
        {
          "kind": "primary",
          "name": "tremor-runtime",
          "url": "https://github.com/tremor-rs/tremor-runtime"
        }
      ],
      "score": {
        "best_practices": 70,
        "documentation": 80,
        "global": 83,
        "license": 80,
        "score_kind": "Primary",
        "security": 100
      },
      "updated_at": 1645603208
    }
  ],
  "Pagination-Total-Count": "7"
}
