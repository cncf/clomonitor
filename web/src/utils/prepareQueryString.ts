import { isEmpty, isUndefined } from 'lodash';

import { BasicQuery, SearchFiltersURL, SearchQuery } from '../types';

export const getURLSearchParams = (query: BasicQuery): URLSearchParams => {
  const q = new URLSearchParams();
  if (!isUndefined(query.filters) && !isEmpty(query.filters)) {
    Object.keys(query.filters).forEach((filterId: string) => {
      return query.filters![filterId].forEach((id: string | number) => {
        q.append(filterId, id.toString());
      });
    });
  }
  if (!isUndefined(query.text)) {
    q.set('text', query.text);
  }
  return q;
};

export const prepareAPIQueryString = (query: SearchQuery): string => {
  const q = getURLSearchParams(query);
  q.set('limit', query.limit.toString());
  q.set('offset', query.offset.toString());
  return `?${q.toString()}`;
};

export const prepareQueryString = (query: SearchFiltersURL): string => {
  const q = getURLSearchParams(query);
  q.set('page', query.pageNumber.toString());
  return `?${q.toString()}`;
};
