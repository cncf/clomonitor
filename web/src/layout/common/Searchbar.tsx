import { isNull } from 'lodash';
import { ChangeEvent, KeyboardEvent, useRef, useState } from 'react';
import { FiSearch } from 'react-icons/fi';
import { useNavigate } from 'react-router-dom';

import { prepareQueryString } from '../../utils/prepareQueryString';
import styles from './Searchbar.module.css';

interface Props {
  classNameWrapper?: string;
}

const Searchbar = (props: Props) => {
  const navigate = useNavigate();
  const inputEl = useRef<HTMLInputElement>(null);
  const [value, setValue] = useState<string>('');

  const onKeyDown = (e: KeyboardEvent<HTMLInputElement>): void => {
    if (e.key === 'Enter') {
      navigate({
        pathname: '/search',
        search: prepareQueryString({
          pageNumber: 1,
          text: value,
          filters: {},
        }),
      });
      setValue('');
      forceBlur();
    }
  };

  const forceBlur = (): void => {
    if (!isNull(inputEl) && !isNull(inputEl.current)) {
      inputEl.current.blur();
    }
  };

  return (
    <div className={`position-relative ${props.classNameWrapper}`}>
      <div
        className={`d-flex align-items-center overflow-hidden searchBar lh-base bg-white mx-auto ${styles.searchBar} search`}
      >
        <div className={`d-flex align-items-center ${styles.iconWrapper} lightIcon`}>
          <FiSearch />
        </div>

        <input
          ref={inputEl}
          className={`flex-grow-1 ps-2 ps-md-0 border-0 shadow-none bg-transparent lh-base ${styles.input}`}
          type="text"
          value={value}
          autoComplete="off"
          autoCorrect="off"
          autoCapitalize="none"
          spellCheck="false"
          placeholder="Search projects"
          onKeyDown={onKeyDown}
          onChange={(e: ChangeEvent<HTMLInputElement>) => setValue(e.target.value)}
        />
      </div>
    </div>
  );
};

export default Searchbar;
