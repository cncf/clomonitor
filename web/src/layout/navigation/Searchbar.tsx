import classNames from 'classnames';
import { isNull } from 'lodash';
import { ChangeEvent, Dispatch, KeyboardEvent, SetStateAction, useEffect, useRef, useState } from 'react';
import { FiSearch } from 'react-icons/fi';
import { IoCloseSharp } from 'react-icons/io5';
import { useNavigate, useSearchParams } from 'react-router-dom';

import API from '../../api';
import { DEFAULT_SORT_BY, DEFAULT_SORT_DIRECTION } from '../../data';
import useBreakpointDetect from '../../hooks/useBreakpointDetect';
import useOutsideClick from '../../hooks/useOutsideClick';
import { Project } from '../../types';
import prepareQueryString from '../../utils/prepareQueryString';
import FoundationBadge from '../common/badges/FoundationBadge';
import HoverableItem from '../common/HoverableItem';
import Image from '../common/Image';
import RoundScore from '../common/RoundScore';
import styles from './Searchbar.module.css';

interface Props {
  setScrollPosition: Dispatch<SetStateAction<number | undefined>>;
  classNameWrapper?: string;
}

const SEARCH_DELAY = 3 * 100; // 300ms
const MIN_CHARACTERS_SEARCH = 2;

const Searchbar = (props: Props) => {
  const navigate = useNavigate();
  const [searchParams] = useSearchParams();
  const inputEl = useRef<HTMLInputElement>(null);
  const dropdownRef = useRef(null);
  const [value, setValue] = useState<string>('');
  const [currentSearch, setCurrentSearch] = useState<string | null>(null);
  const [projects, setProjects] = useState<Project[] | null>(null);
  const [totalProjectsNumber, setTotalProjectsNumber] = useState<number | null>(null);
  const [visibleDropdown, setVisibleDropdown] = useState<boolean>(false);
  const [highlightedItem, setHighlightedItem] = useState<number | null>(null);
  const [dropdownTimeout, setDropdownTimeout] = useState<NodeJS.Timeout | null>(null);
  const point = useBreakpointDetect();

  useOutsideClick([dropdownRef], visibleDropdown, () => {
    cleanProjectsSearch();
  });

  const onKeyDown = (e: KeyboardEvent<HTMLInputElement>): void => {
    switch (e.key) {
      case 'Escape':
        cleanProjectsSearch();
        return;
      case 'ArrowDown':
        updateHighlightedItem('down');
        return;
      case 'ArrowUp':
        updateHighlightedItem('up');
        return;
      case 'Enter':
        e.preventDefault();
        if (!isNull(projects) && !isNull(highlightedItem)) {
          if (highlightedItem === projects.length) {
            search();
          } else {
            const selectedProject = projects[highlightedItem];
            if (selectedProject) {
              goToProject(selectedProject);
            }
          }
        } else {
          search();
        }
        return;
      default:
        return;
    }
  };

  const goToProject = (selectedProject: Project) => {
    forceBlur();
    setValue('');
    cleanProjectsSearch();
    navigate(`/projects/${selectedProject.foundation}/${selectedProject.name}`);
  };

  const forceBlur = (): void => {
    if (!isNull(inputEl) && !isNull(inputEl.current)) {
      inputEl.current.blur();
    }
  };

  const forceFocus = (): void => {
    if (!isNull(inputEl) && !isNull(inputEl.current)) {
      inputEl.current.focus();
    }
  };

  const search = () => {
    props.setScrollPosition(0);
    cleanTimeout();
    cleanProjectsSearch();

    navigate({
      pathname: '/search',
      search: prepareQueryString({
        pageNumber: 1,
        text: value,
        filters: {},
      }),
    });
    forceBlur();
  };

  const cleanTimeout = () => {
    if (!isNull(dropdownTimeout)) {
      clearTimeout(dropdownTimeout);
      setDropdownTimeout(null);
    }
  };

  const cleanSearchValue = () => {
    if (currentSearch === value) {
      props.setScrollPosition(0);
      navigate({
        pathname: '/search',
        search: prepareQueryString({
          pageNumber: 1,
          text: '',
          filters: {},
        }),
      });
    } else {
      setValue('');
    }
    forceFocus();
  };

  const cleanProjectsSearch = () => {
    setProjects(null);
    setTotalProjectsNumber(null);
    setVisibleDropdown(false);
    setHighlightedItem(null);
  };

  const updateHighlightedItem = (arrow: 'up' | 'down') => {
    if (!isNull(projects) && visibleDropdown) {
      if (!isNull(highlightedItem)) {
        let newIndex: number = arrow === 'up' ? highlightedItem - 1 : highlightedItem + 1;
        if (newIndex > projects.length) {
          newIndex = 0;
        }
        if (newIndex < 0) {
          newIndex = projects.length;
        }
        setHighlightedItem(newIndex);
      } else {
        if (projects && projects.length > 0) {
          const newIndex = arrow === 'up' ? projects.length : 0; // We don't subtract 1 because See all results (x) has to be count
          setHighlightedItem(newIndex);
        }
      }
    }
  };

  async function searchProjects() {
    try {
      const searchResults = await API.searchProjects({
        limit: 5,
        offset: 0,
        text: value,
        sort_by: DEFAULT_SORT_BY,
        sort_direction: DEFAULT_SORT_DIRECTION,
      });
      const total = parseInt(searchResults['Pagination-Total-Count']);
      if (total > 0) {
        const isInputFocused = inputEl.current === document.activeElement;
        // We have to be sure that input has focus to display results
        if (isInputFocused) {
          setProjects(searchResults.items);
          setTotalProjectsNumber(total);
          setVisibleDropdown(true);
        } else {
          cleanProjectsSearch();
        }
      } else {
        cleanProjectsSearch();
      }
    } catch (err: any) {
      cleanProjectsSearch();
    }
  }

  useEffect(() => {
    const text = searchParams.get('text');
    setValue(text || '');
    setCurrentSearch(text);
  }, [searchParams]);

  useEffect(() => {
    // Don't display search options for mobile devices
    if (point !== 'xs') {
      const isInputFocused = inputEl.current === document.activeElement;
      if (value.length >= MIN_CHARACTERS_SEARCH && isInputFocused) {
        cleanTimeout();
        setDropdownTimeout(
          setTimeout(() => {
            setHighlightedItem(null);
            searchProjects();
          }, SEARCH_DELAY)
        );
      } else {
        cleanProjectsSearch();
      }
    }

    return () => {
      if (!isNull(dropdownTimeout)) {
        clearTimeout(dropdownTimeout);
      }
    };
  }, [value]); /* eslint-disable-line react-hooks/exhaustive-deps */

  return (
    <div className={`position-relative ${props.classNameWrapper}`}>
      <div
        className={`d-flex align-items-center overflow-hidden searchBar lh-base bg-white mx-auto ${styles.searchBar} search`}
      >
        <input
          data-testid="search-bar"
          ref={inputEl}
          className={`flex-grow-1 ps-2 ps-md-3 border-0 shadow-none bg-transparent lh-base ${styles.input}`}
          type="text"
          value={value}
          autoComplete="off"
          autoCorrect="off"
          autoCapitalize="none"
          spellCheck="false"
          placeholder="Search projects"
          onKeyDown={onKeyDown}
          onChange={(e: ChangeEvent<HTMLInputElement>) => setValue(e.target.value)}
        />

        {value !== '' && (
          <>
            <button
              aria-label="Clear search"
              className={`btn btn-link lh-1 px-2 ${styles.btnIcon}`}
              onClick={cleanSearchValue}
            >
              <div className="text-muted lightIcon">
                <IoCloseSharp />
              </div>
            </button>
            <div className={`vr ${styles.vr}`} />
          </>
        )}

        <button aria-label="Search text" className={`btn btn-link lh-1 px-2 ${styles.btnIcon}`} onClick={search}>
          <div className={`${styles.iconWrapper} lightIcon`}>
            <FiSearch />
          </div>
        </button>
      </div>

      {visibleDropdown && !isNull(projects) && (
        <div
          ref={dropdownRef}
          className={`dropdown-menu dropdown-menu-left p-0 shadow-sm w-100 rounded-0 show noFocus ${styles.dropdown}`}
          role="listbox"
          id="search-list"
        >
          <HoverableItem onLeave={() => setHighlightedItem(null)}>
            <>
              {projects.map((project: Project, index: number) => {
                return (
                  <HoverableItem
                    key={`pkg_${project.id}`}
                    onHover={() => setHighlightedItem(index)}
                    onLeave={() => setHighlightedItem(null)}
                  >
                    <button
                      type="button"
                      className={classNames(
                        'btn btn-link w-100 border-bottom rounded-0 d-flex flex-row align-items-stretch text-decoration-none text-dark p-3',
                        styles.btnProject,
                        { [styles.activeDropdownItem]: index === highlightedItem }
                      )}
                      onClick={() => {
                        goToProject(project);
                      }}
                      aria-label={`Open package ${project.display_name || project.name} detail`}
                      role="option"
                      aria-selected={index === highlightedItem}
                      id={`sl-opt${index}`}
                    >
                      <div className="d-flex flex-row align-items-center w-100">
                        <div
                          className={`d-flex align-items-center justify-content-center me-2 ${styles.miniImageWrapper}`}
                        >
                          <Image alt={`${project.name}`} url={project.logo_url} dark_url={project.logo_dark_url} />
                        </div>
                        <div className="flex-grow-1 d-flex flex-column w-100 truncateWrapper">
                          <div className="d-flex flex-row justify-content-between align-items-end">
                            <span className={`text-truncate fw-bold mb-0 ${styles.title}`}>
                              {project.display_name || project.name}
                            </span>
                          </div>

                          <div className="d-flex flex-row align-items-center mt-1">
                            <FoundationBadge foundation={project.foundation} />
                          </div>
                        </div>

                        <div>
                          <RoundScore score={project.score.global!} className={`ms-2 ${styles.global}`} />
                        </div>
                      </div>
                    </button>
                  </HoverableItem>
                );
              })}

              <HoverableItem
                onHover={() => setHighlightedItem(projects.length)}
                onLeave={() => setHighlightedItem(null)}
              >
                <button
                  type="button"
                  className={classNames('btn btn-link w-100 text-dark p-2', styles.dropdownItem, {
                    [styles.activeDropdownItem]: projects.length === highlightedItem,
                  })}
                  onClick={search}
                  aria-label="See all results"
                  role="option"
                  aria-selected={projects.length === highlightedItem}
                  id={`sl-opt${projects.length}`}
                >
                  See all results ({totalProjectsNumber})
                </button>
              </HoverableItem>
            </>
          </HoverableItem>
        </div>
      )}
    </div>
  );
};

export default Searchbar;
