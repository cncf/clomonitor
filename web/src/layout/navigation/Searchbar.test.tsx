import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { mocked } from 'jest-mock';
import ReactRouter, { BrowserRouter as Router } from 'react-router-dom';

import API from '../../api';
import { Project } from '../../types';
import prepareQueryString from '../../utils/prepareQueryString';
import Searchbar from './Searchbar';
jest.mock('../../api');

const mockUseNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
  ...(jest.requireActual('react-router-dom') as any),
  useNavigate: () => mockUseNavigate,
  useSearchParams: jest.fn(),
}));

interface SearchResults {
  items: Project[];
  'Pagination-Total-Count': string;
}

const getMockSearch = (fixtureId: string): SearchResults => {
  return require(`./__fixtures__/Searchbar/${fixtureId}.json`) as SearchResults;
};

const mockSetScrollPosition = jest.fn();

const defaultProps = {
  setScrollPosition: mockSetScrollPosition,
};

describe('Searchbar', () => {
  beforeEach(() => {
    jest.spyOn(ReactRouter, 'useSearchParams').mockImplementation(() => [new URLSearchParams(''), jest.fn()]);
  });

  afterEach(() => {
    jest.resetAllMocks();
  });

  it('creates snapshot', () => {
    const { asFragment } = render(
      <Router>
        <Searchbar {...defaultProps} />
      </Router>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders proper content', () => {
    render(
      <Router>
        <Searchbar {...defaultProps} />
      </Router>
    );

    expect(screen.getByPlaceholderText('Search projects')).toBeInTheDocument();
    expect(screen.queryByRole('button', { name: 'Clear search' })).toBeNull();
    expect(screen.getByRole('button', { name: 'Search text' })).toBeInTheDocument();
  });

  it('renders with text', () => {
    jest.spyOn(ReactRouter, 'useSearchParams').mockImplementation(() => [new URLSearchParams('?text=test'), jest.fn()]);

    render(
      <Router>
        <Searchbar {...defaultProps} />
      </Router>
    );

    expect(screen.getByRole('textbox')).toBeInTheDocument();
    expect(screen.getByRole('textbox')).toHaveValue('test');
    expect(screen.getByRole('button', { name: 'Clear search' })).toBeInTheDocument();
    expect(screen.getByRole('button', { name: 'Search text' })).toBeInTheDocument();
  });

  describe('clear btn', () => {
    it('clear input', () => {
      jest
        .spyOn(ReactRouter, 'useSearchParams')
        .mockImplementation(() => [new URLSearchParams(''), jest.fn()])
        .mockImplementationOnce(() => [new URLSearchParams('?text=test'), jest.fn()])
        .mockImplementationOnce(() => [new URLSearchParams('?text=test'), jest.fn()])
        .mockImplementationOnce(() => [new URLSearchParams('?text=test'), jest.fn()])
        .mockImplementationOnce(() => [new URLSearchParams('?text=testing'), jest.fn()])
        .mockImplementationOnce(() => [new URLSearchParams('?text=testing'), jest.fn()]);

      render(
        <Router>
          <Searchbar {...defaultProps} />
        </Router>
      );

      const clearBtn = screen.getByRole('button', { name: 'Clear search' });
      const input = screen.getByRole('textbox');

      expect(input).toBeInTheDocument();
      expect(input).toHaveValue('test');
      userEvent.type(input, 'ing');
      userEvent.click(clearBtn);
      expect(input).toHaveValue('');
    });
  });

  it('updates value on change input', async () => {
    jest
      .spyOn(ReactRouter, 'useSearchParams')
      .mockImplementation(() => [new URLSearchParams('?text=testing'), jest.fn()])
      .mockImplementationOnce(() => [new URLSearchParams(''), jest.fn()]);

    const mockSearch = getMockSearch('1');
    mocked(API).searchProjects.mockResolvedValue(mockSearch);

    render(
      <Router>
        <Searchbar {...defaultProps} />
      </Router>
    );

    const input = screen.getByRole('textbox');

    userEvent.type(input, 'testing');

    expect(input).toHaveValue('testing');

    await waitFor(() => {
      expect(API.searchProjects).toHaveBeenCalledTimes(1);
    });
  });

  describe('search projects', () => {
    it('display search results', async () => {
      jest
        .spyOn(ReactRouter, 'useSearchParams')
        .mockImplementation(() => [new URLSearchParams('?text=testing'), jest.fn()])
        .mockImplementationOnce(() => [new URLSearchParams(''), jest.fn()]);

      const mockSearch = getMockSearch('1');
      mocked(API).searchProjects.mockResolvedValue(mockSearch);

      render(
        <Router>
          <Searchbar {...defaultProps} />
        </Router>
      );

      const input = screen.getByRole('textbox') as HTMLInputElement;

      input.focus();

      userEvent.type(input, 'testing');
      expect(input.value).toBe('testing');

      await waitFor(() => {
        expect(API.searchProjects).toHaveBeenCalledTimes(1);
        expect(API.searchProjects).toHaveBeenCalledWith({
          limit: 5,
          offset: 0,
          sort_by: 'name',
          sort_direction: 'asc',
          text: 'testing',
        });
      });

      expect(screen.getByRole('listbox')).toBeInTheDocument();
      expect(screen.getAllByRole('option')).toHaveLength(8);
    });

    it("doesn't display results when input is not focused", async () => {
      const useSearchParamsSpy = jest.spyOn(ReactRouter, 'useSearchParams');
      useSearchParamsSpy.mockImplementation(() => [new URLSearchParams('?text=test'), jest.fn()]);

      const mockSearch = getMockSearch('1');
      mocked(API).searchProjects.mockResolvedValue(mockSearch);

      render(
        <Router>
          <Searchbar {...defaultProps} />
        </Router>
      );

      const input = screen.getByDisplayValue('test');

      input.focus();
      userEvent.type(input, 'ing');
      input.blur();

      await waitFor(() => {
        expect(API.searchProjects).toHaveBeenCalledTimes(1);
        input.blur();
      });

      expect(screen.queryByRole('listbox')).toBeNull();
    });

    it('loads project detail from search dropdown', async () => {
      jest
        .spyOn(ReactRouter, 'useSearchParams')
        .mockImplementation(() => [new URLSearchParams('?text=testing'), jest.fn()])
        .mockImplementationOnce(() => [new URLSearchParams(''), jest.fn()]);

      const mockSearch = getMockSearch('1');
      mocked(API).searchProjects.mockResolvedValue(mockSearch);

      render(
        <Router>
          <Searchbar {...defaultProps} />
        </Router>
      );

      const input = screen.getByRole('textbox') as HTMLInputElement;

      input.focus();

      userEvent.type(input, 'testing');
      expect(input.value).toBe('testing');

      await waitFor(() => {
        expect(API.searchProjects).toHaveBeenCalledTimes(1);
        expect(API.searchProjects).toHaveBeenCalledWith({
          limit: 5,
          offset: 0,
          sort_by: 'name',
          sort_direction: 'asc',
          text: 'testing',
        });
      });

      expect(screen.getByRole('listbox')).toBeInTheDocument();
      const items = screen.getAllByRole('option');
      userEvent.click(items[1]);

      expect(mockUseNavigate).toHaveBeenCalledTimes(1);
      expect(mockUseNavigate).toHaveBeenCalledWith('/projects/backstage/backstage');
    });

    it('loads new search from search dropdown', async () => {
      jest
        .spyOn(ReactRouter, 'useSearchParams')
        .mockImplementation(() => [new URLSearchParams('?text=testing'), jest.fn()])
        .mockImplementationOnce(() => [new URLSearchParams(''), jest.fn()]);

      const mockSearch = getMockSearch('1');
      mocked(API).searchProjects.mockResolvedValue(mockSearch);

      render(
        <Router>
          <Searchbar {...defaultProps} />
        </Router>
      );

      const input = screen.getByRole('textbox') as HTMLInputElement;

      input.focus();

      userEvent.type(input, 'testing');
      expect(input.value).toBe('testing');

      await waitFor(() => {
        expect(API.searchProjects).toHaveBeenCalledTimes(1);
        expect(API.searchProjects).toHaveBeenCalledWith({
          limit: 5,
          offset: 0,
          sort_by: 'name',
          sort_direction: 'asc',
          text: 'testing',
        });
      });

      expect(screen.getByRole('listbox')).toBeInTheDocument();
      const allResults = screen.getByRole('option', { name: 'See all results' });
      userEvent.click(allResults);

      expect(mockUseNavigate).toHaveBeenCalledTimes(1);
      expect(mockUseNavigate).toHaveBeenCalledWith({ pathname: '/search', search: '?text=testing&page=1' });
    });

    it('uses arrow for seleting one item and loads detail to click enter', async () => {
      jest
        .spyOn(ReactRouter, 'useSearchParams')
        .mockImplementation(() => [new URLSearchParams('?text=testing'), jest.fn()])
        .mockImplementationOnce(() => [new URLSearchParams(''), jest.fn()]);

      const mockSearch = getMockSearch('1');
      mocked(API).searchProjects.mockResolvedValue(mockSearch);

      render(
        <Router>
          <Searchbar {...defaultProps} />
        </Router>
      );

      const input = screen.getByRole('textbox') as HTMLInputElement;

      input.focus();

      userEvent.type(input, 'testing');
      expect(input.value).toBe('testing');

      await waitFor(() => {
        expect(API.searchProjects).toHaveBeenCalledTimes(1);
        expect(API.searchProjects).toHaveBeenCalledWith({
          limit: 5,
          offset: 0,
          sort_by: 'name',
          sort_direction: 'asc',
          text: 'testing',
        });
      });

      expect(screen.getByRole('listbox')).toBeInTheDocument();
      const options = screen.getAllByRole('option');
      expect(options[0]).not.toHaveClass('activeDropdownItem');

      userEvent.keyboard('{arrowdown}');

      await waitFor(() => {
        expect(options[0]).toHaveClass('activeDropdownItem');
      });

      userEvent.keyboard('{arrowdown}{arrowdown}');

      expect(options[0]).not.toHaveClass('activeDropdownItem');
      expect(options[2]).toHaveClass('activeDropdownItem');

      userEvent.keyboard('{enter}');

      expect(mockUseNavigate).toHaveBeenCalledTimes(1);
      expect(mockUseNavigate).toHaveBeenCalledWith('/projects/keptn/keptn');
    });
  });

  describe('Navigate', () => {
    it('calls on Enter key press', () => {
      jest
        .spyOn(ReactRouter, 'useSearchParams')
        .mockImplementation(() => [new URLSearchParams('?text=testing'), jest.fn()])
        .mockImplementationOnce(() => [new URLSearchParams(''), jest.fn()]);

      render(
        <Router>
          <Searchbar {...defaultProps} />
        </Router>
      );

      const input = screen.getByRole('textbox');
      userEvent.type(input, 'testing{enter}');
      expect(input).not.toBe(document.activeElement);
      expect(mockUseNavigate).toHaveBeenCalledTimes(1);
      expect(mockUseNavigate).toHaveBeenCalledWith({
        pathname: '/search',
        search: prepareQueryString({
          text: 'testing',
          pageNumber: 1,
        }),
      });
    });

    it('calls navigate on Enter key press when text is empty with undefined text', () => {
      render(
        <Router>
          <Searchbar {...defaultProps} />
        </Router>
      );

      const input = screen.getByPlaceholderText('Search projects');
      userEvent.type(input, '{enter}');
      expect(mockUseNavigate).toHaveBeenCalledTimes(1);
      expect(mockUseNavigate).toHaveBeenCalledWith({
        pathname: '/search',
        search: prepareQueryString({
          text: undefined,
          pageNumber: 1,
        }),
      });
    });
  });
});
