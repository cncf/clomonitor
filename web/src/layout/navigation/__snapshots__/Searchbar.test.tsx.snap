// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Searchbar creates snapshot 1`] = `
<DocumentFragment>
  <div
    class="position-relative undefined"
  >
    <div
      class="d-flex align-items-center overflow-hidden searchBar lh-base bg-white mx-auto searchBar search"
    >
      <input
        autocapitalize="none"
        autocomplete="off"
        autocorrect="off"
        class="flex-grow-1 ps-2 ps-md-3 border-0 shadow-none bg-transparent lh-base input"
        data-testid="search-bar"
        placeholder="Search projects"
        spellcheck="false"
        type="text"
        value=""
      />
      <button
        aria-label="Search text"
        class="btn btn-link lh-1 px-2 btnIcon"
      >
        <div
          class="iconWrapper lightIcon"
        >
          <svg
            fill="none"
            height="1em"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <circle
              cx="11"
              cy="11"
              r="8"
            />
            <line
              x1="21"
              x2="16.65"
              y1="21"
              y2="16.65"
            />
          </svg>
        </div>
      </button>
    </div>
  </div>
</DocumentFragment>
`;
